## DSA4262 Individual Assignment 1 (Data Cleaing + Visualization Code)

### **Importing Relevant Modules** 
1. **Pandas** - For data cleaning and transformation
2. **Numpy** - For data manipulation 
3. **Matplotlib** - For data visualization

We will not use other data visualization tools such as **plotly** for graph simplicity, as well as refraining from using dynamic plots.

In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 

### **Importing The Needed Datasets**

There are several datasets that will used in this Data Visualization project. These include: 

1. Depression Rates By Country From 2019 To 2023
2. Depression DALY Contribution By Age Group And Gender
3. TBC
4. TBC
5. TBC
6. Singapore Resident Population By Age Group And Gender

All the datasets are in Excel Workbook (.xlsx) and we can import these datasets via the pandas pd.read_excel method.

#### **Importing Dataset 1 - Depression Rates By Country From 2019 To 2023**

In [13]:
depression_rates_country = pd.read_excel("Depression Rates By Country 2019 To 2023.xlsx")
depression_rates_country.head()

flagCode         country  DepressiveDisorderByCountryRate_2023  \
0       SY           Syria                                8.4422   
1       GB  United Kingdom                                7.4354   
2       NL     Netherlands                                7.2209   
3       UA         Ukraine                                6.7550   
4       TN         Tunisia                                6.3499   

   DepressiveDisorderByCountryRate_2022  DepressiveDisorderByCountryRate_2021  \
0                                8.1838                                8.3835   
1                                7.3396                                7.1480   
2                                6.8113                                7.0112   
3                                6.6077                                6.6706   
4                                6.1626                                6.4065   

   DepressiveDisorderByCountryRate_2020  DepressiveDisorderByCountryRate_2019  \
0                                8.1806                                6.5837   
1                                7.1750                                5.8540   
2                                6.9441                                5.7450   
3                                6.4114                                5.3946   
4                                6.2386                                5.0819   

   DepressiveDisorderByCountryPercentage_2023  \
0                                         NaN   
1                                      7.4354   
2                                      7.2209   
3                                      6.7550   
4                                      6.3499   

   DepressiveDisorderByCountryPercentage_2022  \
0                                         NaN   
1                                      7.3396   
2                                      6.8113   
3                                      6.6077   
4                                      6.1626   

   DepressiveDisorderByCountryPercentage_2021  \
0                                         NaN   
1                                      7.1480   
2                                      7.0112   
3                                      6.6706   
4                                      6.4065   

   DepressiveDisorderByCountryPercentage_2020  \
0                                         NaN   
1                                      7.1750   
2                                      6.9441   
3                                      6.4114   
4                                      6.2386   

   DepressiveDisorderByCountryPercentage_2019  \
0                                         NaN   
1                                      5.8540   
2                                      5.7450   
3                                      5.3946   
4                                      5.0819   

   DepressiveDisorderByCountryCases_2023  \
0                              1464840.0   
1                              4867643.0   
2                              1226914.0   
3                              2580229.0   
4                               732908.4   

   DepressiveDisorderByCountryCases_2022  \
0                              1404486.0   
1                              4781675.0   
2                              1147441.0   
3                              2727005.0   
4                               708113.8   

   DepressiveDisorderByCountryCases_2021  \
0                              1437534.0   
1                              4621701.0   
2                              1171730.0   
3                              2773255.0   
4                               732554.9   

   DepressiveDisorderByCountryCases_2020  \
0                              1390187.0   
1                              4606637.0   
2                              1154372.0   
3                              2689636.0   
4                               709482.8   

   DepressiveDisorderByCountryCases_2019  
0                              1116854.0 

#### **Importing Dataset 2 - Depression DALY Contribution By Age Group And Gender**

In [18]:
daly_contribution = pd.read_excel("Depression DALY Contribution By Age Group And Gender (Global).xlsx")
daly_contribution.head()

population_group_id population_group_name  measure_id  \
0                    1        All Population           2   
1                    1        All Population           2   
2                    1        All Population           2   
3                    1        All Population           2   
4                    1        All Population           2   

                             measure_name  location_id location_name  sex_id  \
0  DALYs (Disability-Adjusted Life Years)            1        Global       1   
1  DALYs (Disability-Adjusted Life Years)            1        Global       2   
2  DALYs (Disability-Adjusted Life Years)            1        Global       1   
3  DALYs (Disability-Adjusted Life Years)            1        Global       2   
4  DALYs (Disability-Adjusted Life Years)            1        Global       1   

  sex_name  age_id     age_name  cause_id            cause_name  metric_id  \
0     Male       1     <5 years       567  Depressive disorders          2   
1   Female       1     <5 years       567  Depressive disorders          2   
2     Male       6    5-9 years       567  Depressive disorders          2   
3   Female       6    5-9 years       567  Depressive disorders          2   
4     Male       7  10-14 years       567  Depressive disorders          2   

  metric_name  year           val         upper         lower  
0     Percent  2014  4.694402e-07  8.444146e-07  1.704511e-07  
1     Percent  2014  5.988374e-07  1.063922e-06  1.979924e-07  
2     Percent  2014  2.316760e-03  4.118151e-03  1.329223e-03  
3     Percent  2014  4.730082e-03  8.069187e-03  2.795667e-03  
4     Percent  2014  2.020621e-02  2.873852e-02  1.387388e-02

#### **Importing Dataset 6 - Singapore Resident Population By Age Group And Gender**

In [21]:
singapore_resident_data = pd.read_excel("Singapore Resident Population By Age Group And Gender.xlsx")
singapore_resident_data.head()

Number Total_Total Total_0_4 Total_5_9 Total_10_14  \
0                   Total     3902690    183580    204450      214390   
1       Ang Mo Kio- Total      174770      6790      7660        8290   
2  Ang Mo Kio Town Centre        5020       260       280         320   
3               Cheng San       29770      1290      1180        1290   
4              Chong Boon       27900       910      1100        1180   

  Total_15_19 Total_20_24 Total_25_29 Total_30_34 Total_35_39  ...  \
0      242900      264130      271030      290620      301070  ...   
1        9320       10310       11170       12250       13070  ...   
2         280         260         310         370         420  ...   
3        1400        1570        1830        2490        2490  ...   
4        1370        1520        1800        1980        2100  ...   

  Females_40_44 Females_45_49 Females_50_54 Females_55_59 Females_60_64  \
0        162300        153810        156630        147200        120830   
1          7100          6660          7010          7120          6940   
2           260           210           170           170           160   
3          1260          1110          1140          1220          1270   
4          1060          1080          1120          1160          1200   

  Females_65_69 Females_70_74 Females_75_79 Females_80_84 Females_85andOver  
0         93730         54850         45090         30850             27660  
1          5810          3600          2800          1910              1520  
2           150            80            70            40                40  
3           970           580           420           280               260  
4          1050           690           500           310               240  

[5 rows x 58 columns]

--------------------------------------------------------------------------------------------------------------------------

### **Data Cleaning**

We will need to perform data cleaning from the raw data above. The data cleaning methods will include (not exhaustive). 

1) Removing unnecessary and irrelevant columns 
2) Renaming columns to a more readable format 
3) Imputation of missing (NaN) values 
4) Removal of unnecessary rows (specifically for the Singapore dataset)
5) Standardizing Age Group Categories (<5 years, 5 to 9 years, 10 to 14 years etc.)

#### **Data Cleaning - Dataset 1 (Depression Rates By Country From 2019 To 2023)**

Data Cleaning Step Required For Dataset 1: 

*   Removing unnecessary columns (flagCode, Rate Columns, Cases Columns)
*   Renaming useful columns to more readable formats
*   Handling missing values (NaN) for useful columns 
*   Converting all values to 2 decimal places for consistency
*   Add an additional column called "Continent" to place the countries according to continent (for visualization purposes later)

##### **Step 1: Removing Unnecessary Columns**

Unnecessary columns include 

*   **flagCode** -> We can directly use the Country column instead
*   **Rate Columns** -> Rate value calculations involve complex metrics and this not suitable for explaining
*   **Cases Columns** -> Number of cases is not an accurate metric, different countries have different populations and it is obvious that a country with very high population (ega Chins, India, US) have much higher number of cases than a country with very low population (eg. Singapore, Bhutan, San Marino). It is not a fair comparison.

We will only keep useful columns such as country and Percentage columns in the dataset. 

In [33]:
# Keeping only the relevant columns
depression_rates_country = depression_rates_country[['country', 'DepressiveDisorderByCountryPercentage_2023', 
                                                     'DepressiveDisorderByCountryPercentage_2022',
                                                     'DepressiveDisorderByCountryPercentage_2021',
                                                     'DepressiveDisorderByCountryPercentage_2020',
                                                     'DepressiveDisorderByCountryPercentage_2019']]

depression_rates_country.head() 

country  DepressiveDisorderByCountryPercentage_2023  \
0           Syria                                         NaN   
1  United Kingdom                                      7.4354   
2     Netherlands                                      7.2209   
3         Ukraine                                      6.7550   
4         Tunisia                                      6.3499   

   DepressiveDisorderByCountryPercentage_2022  \
0                                         NaN   
1                                      7.3396   
2                                      6.8113   
3                                      6.6077   
4                                      6.1626   

   DepressiveDisorderByCountryPercentage_2021  \
0                                         NaN   
1                                      7.1480   
2                                      7.0112   
3                                      6.6706   
4                                      6.4065   

   DepressiveDisorderByCountryPercentage_2020  \
0                                         NaN   
1                                      7.1750   
2                                      6.9441   
3                                      6.4114   
4                                      6.2386   

   DepressiveDisorderByCountryPercentage_2019  
0                                         NaN  
1                                      5.8540  
2                                      5.7450  
3                                      5.3946  
4                                      5.0819

##### **Step 2: Rename Useful Columns To A Readable Format**

For now, the column names DepressiveDisorderByCountryPercentage_2023 is very long and hard to read. We can rename such columns to "Depression Percentage (2023)". This can be applied to all the other year percentage columns from 2019 to 2023. We will also standardize all column names to camel case (eg. country -> Country).

In [36]:
# Renaming the useful columns to a more readable format
depression_rates_country.rename(columns = {'country': 'Country', 
                                           'DepressiveDisorderByCountryPercentage_2023': 'Depression Percentage (2023)',
                                           'DepressiveDisorderByCountryPercentage_2022': 'Depression Percentage (2022)',
                                           'DepressiveDisorderByCountryPercentage_2021': 'Depression Percentage (2021)',
                                           'DepressiveDisorderByCountryPercentage_2020': 'Depression Percentage (2020)',
                                           'DepressiveDisorderByCountryPercentage_2019': 'Depression Percentage (2019)'}, 
                                           inplace = True)

depression_rates_country.head()

Country  Depression Percentage (2023)  Depression Percentage (2022)  \
0           Syria                           NaN                           NaN   
1  United Kingdom                        7.4354                        7.3396   
2     Netherlands                        7.2209                        6.8113   
3         Ukraine                        6.7550                        6.6077   
4         Tunisia                        6.3499                        6.1626   

   Depression Percentage (2021)  Depression Percentage (2020)  \
0                           NaN                           NaN   
1                        7.1480                        7.1750   
2                        7.0112                        6.9441   
3                        6.6706                        6.4114   
4                        6.4065                        6.2386   

   Depression Percentage (2019)  
0                           NaN  
1                        5.8540  
2                        5.7450  
3                        5.3946  
4                        5.0819

##### **Step 3: Handling Missing/NaN Values For Useful Columns**

From the dataset result above, we can see that for Syria, the Depression Percentage columns have NaN values, implying that there is missing data for Syria. Other countries listed such as UK, Netherlands and Ukraine have values. There are a couple of possible ways to handle the missing data. 

1) **Drop the rows with missing data** -> To be used if the proportion of NaN values is small (probably less than 10% of the data). 
2) **Imputation using mean/median** -> To be used if the proportion of NaN values is larger (20% or more) such that dropping these rows will affect the accuracy of the depression percentages in subsequent steps.

We first need to determine the proprtion of NaN values for each column. 

In [39]:
# Determining the Number and Proportion of NaN in the dataset
nan_summary = pd.DataFrame({
    "Number Of NaN": depression_rates_country.isna().sum(),
    "Proportion Of NaN": depression_rates_country.isna().sum() / len(depression_rates_country)
})

print(nan_summary)

                              Number Of NaN  Proportion Of NaN
Country                                   0           0.000000
Depression Percentage (2023)             18           0.089109
Depression Percentage (2022)             18           0.089109
Depression Percentage (2021)             18           0.089109
Depression Percentage (2020)             18           0.089109
Depression Percentage (2019)             18           0.089109


We observe that for the Depression Percentage columns, the proportion of NaN is only 0.089109 (around 9%), which is less than 10% of the data. Hence, we could safely drop the rows with missing data without losing too much information and data. More than 90% of the countries are still present in the dataset. 

In [42]:
# Dropping all rows containing any NaN value
depression_rates_country = depression_rates_country.dropna()
depression_rates_country.head()

Country  Depression Percentage (2023)  Depression Percentage (2022)  \
1  United Kingdom                        7.4354                        7.3396   
2     Netherlands                        7.2209                        6.8113   
3         Ukraine                        6.7550                        6.6077   
4         Tunisia                        6.3499                        6.1626   
5         Lebanon                        6.1181                        5.9142   

   Depression Percentage (2021)  Depression Percentage (2020)  \
1                        7.1480                        7.1750   
2                        7.0112                        6.9441   
3                        6.6706                        6.4114   
4                        6.4065                        6.2386   
5                        6.0345                        5.9202   

   Depression Percentage (2019)  
1                        5.8540  
2                        5.7450  
3                        5.3946  
4                        5.0819  
5                        4.8078

Here, we observe that the row where the country is Syria has been removed from the dataset due to the presence of NaN values. Now, the first row of the dataset is United Kingdom (originally the second row, after Syria). This shows that the rows containing NaN values have been effectively removed from the dataset.

##### **Step 4: Adding An Additional Column, Continent To Classify The Countries**

This step is necessary as there are too many countries (>180 countries) and imagine creating a bar chart or a line graph where there are such number of categories. Grouping the countries by continent makes much more sense (eg. Asia, Africa, North America, South America, Europe, Oceania) as there will only be 6 categories instead of the initial 180+ categories, making subsequent analysis and visualizations more understandable and more user-friendly. 

We will be using another dataset, Country Continent Groupings, from Our World In Data. This dataset contains the name of the country, the country code, as well as the continent the country is in. 

In [46]:
country_continent = pd.read_excel("Country Continent Groupings.xlsx")
country_continent.head()

Entity Code  Year World region according to OWID
0     Afghanistan  AFG  2023                           Asia
1   Aland Islands  ALA  2023                         Europe
2         Albania  ALB  2023                         Europe
3         Algeria  DZA  2023                         Africa
4  American Samoa  ASM  2023                        Oceania

Similarly, we can clean this dataset by 

1) Selecting the relevant columns - Entity (Country) and World region according to OWID (Continent). Year and Code are not useful columns in our analysis and thus can be removed from the original dataset. 

2) Renaming the useful columns: 
*   Entity -> Country
*   World region according to OWID -> Continent



In [49]:
# Selecting The Relevant Columns - Country and Continent
country_continent = country_continent[['Entity', 'World region according to OWID']]

# Renaming the Relevant Columns
country_continent.rename(columns = {'Entity': 'Country', 
                                    'World region according to OWID': 'Continent'},
                                    inplace = True)

country_continent.head()

Country Continent
0     Afghanistan      Asia
1   Aland Islands    Europe
2         Albania    Europe
3         Algeria    Africa
4  American Samoa   Oceania

Afterwards, we can merge the two datasets (country_continent and depression_rates_country) together using the Country column as the column of similarity. We can rely on `pandas pd.merge()` method to combine the two datasets. We will name the merged dataset `depression_rates_continent`.

In [52]:
# Merging The Two Datasets Together (Country Common Column)
depression_rates_continent = pd.merge(depression_rates_country, 
                                      country_continent, 
                                      on = 'Country', 
                                      how = 'left')

# Rearraging The Columns
new_order = ['Country', 'Continent', 'Depression Percentage (2019)', 'Depression Percentage (2020)',
             'Depression Percentage (2021)', 'Depression Percentage (2022)', 'Depression Percentage (2023)']

depression_rates_continent = depression_rates_continent[new_order]
depression_rates_continent.head()

Country Continent  Depression Percentage (2019)  \
0  United Kingdom    Europe                        5.8540   
1     Netherlands    Europe                        5.7450   
2         Ukraine    Europe                        5.3946   
3         Tunisia    Africa                        5.0819   
4         Lebanon      Asia                        4.8078   

   Depression Percentage (2020)  Depression Percentage (2021)  \
0                        7.1750                        7.1480   
1                        6.9441                        7.0112   
2                        6.4114                        6.6706   
3                        6.2386                        6.4065   
4                        5.9202                        6.0345   

   Depression Percentage (2022)  Depression Percentage (2023)  
0                        7.3396                        7.4354  
1                        6.8113                        7.2209  
2                        6.6077                        6.7550  
3                        6.1626                        6.3499  
4                        5.9142                        6.1181

Now, the dataset looks much more readable, user-friendly and clean. However, a problem might happen when joining the two datasets. It is possible that a country's name is different for the two datasets that are merged. For example, the `depression_rates_continent` dataset might state **"United States"** while the `country_continent` dataset might state **"United States Of America"**, leading to no joining. We need to verify if such records exist. For these records, the **Continent** column will be represented by **NaN**.

In [55]:
# Finding NaN rows for Continent
rows_with_continent_nan = depression_rates_continent[depression_rates_continent.isna().any(axis=1)]
rows_with_continent_nan

Country Continent  Depression Percentage (2019)  \
182  Timor-Leste       NaN                        2.0805   

     Depression Percentage (2020)  Depression Percentage (2021)  \
182                        2.3825                        2.5131   

     Depression Percentage (2022)  Depression Percentage (2023)  
182                        2.5865                        2.4806

Indeed, there is one such row - where country is `Timor-Leste`. Since a **Left Join** is being used, the `depression_rates_country` names the country as "Timor-Leste" but the `country_continent` names the country another way. Upon further investigation, the name stated in `country_continent` is "East Timor". The continent of East Timor/Timor-Leste is supposed to be **Asia**. Hence, we can replace the NaN value with Asia.

In [58]:
# Replace the NaN value in the Continent column to Asia
depression_rates_continent["Continent"] = depression_rates_continent["Continent"].fillna("Asia")

depression_rates_continent[depression_rates_continent["Country"] == "Timor-Leste"]

Country Continent  Depression Percentage (2019)  \
182  Timor-Leste      Asia                        2.0805   

     Depression Percentage (2020)  Depression Percentage (2021)  \
182                        2.3825                        2.5131   

     Depression Percentage (2022)  Depression Percentage (2023)  
182                        2.5865                        2.4806

##### **Step 5: Grouping The Countries By Continent and Finding The Average Depression Percentage Rates**

There are now 183 rows in the dataset. If we group by country, there will be too many categories (183 distinct categories) for subsequent analysis and visuslizations. Hence, we will group the countries by continent (6 continents involved) and afterwards, determine the average depression percentage rates for each continent. 

Moreover, we will round each of the average depression rates to 2 decimal places for simplicity, as well as making the visualizations more reader-friendly subsequently.

In [61]:
year_cols = ["Depression Percentage (2019)", "Depression Percentage (2020)",
             "Depression Percentage (2021)", "Depression Percentage (2022)",
             "Depression Percentage (2023)"]

# Grouping the countries by continent and finding the average depression rates
avg_depression_by_continent = (depression_rates_continent.groupby("Continent")[year_cols].mean().reset_index())

# Rounding the average depression rates to 2 decimal places (for simplicity)
avg_depression_by_continent = avg_depression_by_continent.round(2)

avg_depression_by_continent

Continent  Depression Percentage (2019)  Depression Percentage (2020)  \
0         Africa                          3.03                          3.68   
1           Asia                          3.23                          3.82   
2         Europe                          3.63                          4.27   
3  North America                          3.37                          4.07   
4        Oceania                          2.80                          3.28   
5  South America                          3.22                          3.94   

   Depression Percentage (2021)  Depression Percentage (2022)  \
0                          3.75                          3.72   
1                          3.91                          3.84   
2                          4.30                          4.26   
3                          4.19                          4.10   
4                          3.41                          3.35   
5                          4.05                          4.01   

   Depression Percentage (2023)  
0                          3.78  
1                          3.90  
2                          4.34  
3                          4.19  
4                          3.40  
5                          4.05

--------------------------------------------------------------------------------------------------------------------------

#### **Data Cleaning - Dataset 2 (Depression DALY Contribution By Age Group And Gender)**

Data Cleaning Steps Required For Dataset 2:

Obtaining the rows where year is in 2023 (the most recent year in the dataset)
Removing unnecessary columns (ID columns, measure_name etc.)
Renaming useful columns to more readable formats
Multiplying some the column values by 100 to convert them to percentages, rounding them to 2 decimal places for simplicity and consistency
Altering the values of the age range column (if necessary)

##### **Step 1: Extracting Rows Where Year = 2023**

The year column in this dataset ranges from 2014 to 2023. Since we do not need many years to analyze, we should keep the most recent year, 2023, for investigation. Furthermore, the year 2023 tallies with the most recent year in the first dataset (Depression Rates By Country 2019 to 2023) so that we can examine which gender and age group face the most depression subsequently. 

In [66]:
# Extract Rows Where Year = 2023 only
daly_contribution = daly_contribution[daly_contribution["year"] == 2023]
daly_contribution.head()

population_group_id population_group_name  measure_id  \
268                    1        All Population           2   
269                    1        All Population           2   
270                    1        All Population           2   
271                    1        All Population           2   
424                    1        All Population           2   

                               measure_name  location_id location_name  \
268  DALYs (Disability-Adjusted Life Years)            1        Global   
269  DALYs (Disability-Adjusted Life Years)            1        Global   
270  DALYs (Disability-Adjusted Life Years)            1        Global   
271  DALYs (Disability-Adjusted Life Years)            1        Global   
424  DALYs (Disability-Adjusted Life Years)            1        Global   

     sex_id sex_name  age_id     age_name  cause_id            cause_name  \
268       1     Male     232  65-74 years       567  Depressive disorders   
269       2   Female     232  65-74 years       567  Depressive disorders   
270       1     Male     243  75-84 years       567  Depressive disorders   
271       2   Female     243  75-84 years       567  Depressive disorders   
424       1     Male       1     <5 years       567  Depressive disorders   

     metric_id metric_name  year           val     upper         lower  
268          2     Percent  2023  7.784041e-03  0.010710  5.445612e-03  
269          2     Percent  2023  1.544716e-02  0.021064  1.124679e-02  
270          2     Percent  2023  5.113234e-03  0.006974  3.563535e-03  
271          2     Percent  2023  9.556417e-03  0.012812  6.852055e-03  
424          2     Percent  2023  7.246060e-07  0.000002  2.079482e-07

##### **Step 2: Removing Unnecessary Columns**

Unnecessary columns include 

*   **population_group_id**, **measure_id**, **location_id**, **sex_id**, **age_id**, **cause_id**, **metric_id** -> ID numbers which bring no value for analysis
*   **upper**, **lower** -> Simply the upper and lower bounds of the possible values for DALY which make analysis too complex. We should ignore these columns and just focus on the exact DALY values.  
*   **metric_name** -> All rows have metric in percentages which will be revealed in the visualizations subsequently.
*   **year** -> After Step 1, all rows are in Year 2023, which makes this column now redundant. 
*   **population_group_name**, **measure_name**, **location_name**, **cause_name** -> Similar to year, they all have only 1 value and are redundant. 

We will only keep useful columns such as `sex_name`, `age_name` and `val` which are useful for analysis and visualization. 

In [69]:
# Keeping only the relevant columns
daly_contribution = daly_contribution[['sex_name', 'age_name', 'val']]

daly_contribution.head()  

sex_name     age_name           val
268     Male  65-74 years  7.784041e-03
269   Female  65-74 years  1.544716e-02
270     Male  75-84 years  5.113234e-03
271   Female  75-84 years  9.556417e-03
424     Male     <5 years  7.246060e-07

##### **Step 3: Rename Useful Columns To A Readable Format**

For now, the column names sex_name, age_name and val is either confusing or vague. We can rename the columns to a more understandable and clear names, such as: 

*   sex_name -> Gender
*   age_name -> Age Category
*   val -> Depression DALY Percntage


In [72]:
# Renaming the useful columns to a more readable format
daly_contribution.rename(columns = {'sex_name': 'Gender', 
                                    'age_name': 'Age Category',
                                    'val': 'Depression DALY Percentage'}, 
                                    inplace = True)

daly_contribution.head()

Gender Age Category  Depression DALY Percentage
268    Male  65-74 years                7.784041e-03
269  Female  65-74 years                1.544716e-02
270    Male  75-84 years                5.113234e-03
271  Female  75-84 years                9.556417e-03
424    Male     <5 years                7.246060e-07

##### **Step 4: Multiplying Percentage Values By 100 For Accuracy**

For now, the values in the `Depression DALY Percentage` column are still in decimals. We need to multiply all values in the column by 100 to convert them to percentages. This will prevent confusion in the readers. 

Afterwards, we can convert each of the percentage values to 2 decimal places for simplcity.

In [75]:
# Multiplying all values in Depression DALY Percentage by 100 to convert to percentages
daly_contribution["Depression DALY Percentage"] = daly_contribution["Depression DALY Percentage"] * 100

# Converting the peecentages to 2 decimal places
daly_contribution["Depression DALY Percentage"] = daly_contribution["Depression DALY Percentage"].round(2)

daly_contribution.head()

Gender Age Category  Depression DALY Percentage
268    Male  65-74 years                        0.78
269  Female  65-74 years                        1.54
270    Male  75-84 years                        0.51
271  Female  75-84 years                        0.96
424    Male     <5 years                        0.00

##### **Step 5: Exploring The Age Category Column In Dataset**

We need to explore the unique categories of the Age Category dataset. (There is actually a problem with some category values in this column!) We can achieve this by using the `.unique()` method.

In [78]:
# Finding the distinct values in the Age Category column
daly_contribution["Age Category"].unique()

array(['65-74 years', '75-84 years', '<5 years', '5-9 years',
       '10-14 years', '15-19 years', '20-24 years', '25-29 years',
       '30-34 years', '35-39 years', '40-44 years', '45-49 years',
       '50-54 years', '55-59 years', '60-64 years', '65-69 years',
       '70-74 years', '75-79 years', '80-84 years', '85-89 years',
       '90-94 years'], dtype=object)

From the results obtained, we can see that there is some overlap between some of the values. This is not ideal and some categories should be removed to ensure accuracy. 

For example, there is a category called **'65-74 years'** and there are two other categories **'65-69 years'** and **'70-74 years'** which are overlapping. Another example will be **'75-84 years'**, **'75-79 years'** and **'80-84 years'**. 

Since most of the categories are in intervals of 5 years (eg. '20-24 years', where the oldest is 24 years old and the youngest is 20 years old), we should keep all 5-year interval ranges and discard the 10-year interval ranges for consistency.

In [81]:
# Removing Rows With Age Category of 10-year intervals
daly_contribution = daly_contribution[~daly_contribution["Age Category"].isin(["65-74 years", "75-84 years"])]

daly_contribution.head()

Gender Age Category  Depression DALY Percentage
424    Male     <5 years                        0.00
425  Female     <5 years                        0.00
426    Male    5-9 years                        0.27
427  Female    5-9 years                        0.54
428    Male  10-14 years                        2.39

Lastly, we will remove the word **'years'** from the Age Category column to display only the range in numbers for ease of reading. For instance, **"5-9 years"** will become **"5-9"** without the word years. 

In [84]:
# Remove the 'years' word from the Age Category column
daly_contribution["Age Category"] = daly_contribution["Age Category"].str.replace(" years", "", regex = False)

daly_contribution.head()

Gender Age Category  Depression DALY Percentage
424    Male           <5                        0.00
425  Female           <5                        0.00
426    Male          5-9                        0.27
427  Female          5-9                        0.54
428    Male        10-14                        2.39

----------------------------------------------------------------------------------------------------------------------------

#### **Dataset Cleaing - Dataset 3 (Singapore Resident Population By Age Group And Gender)**

Data Cleaning Step Required For Dataset 3: 

*   Removing unnecessary rows (districts in estates, grand total etc.)
*   Handling NA values
*   Converting absolute numbers to proportion
*   Removing unnecessary columns (Total_Total)
*   Renaming useful columns to more readable and understandable formats


##### **Step 1: Removing Unnecessary Rows**

There are now many rows in the dataset, but some are districts while others are regions. As there are too many regions in Singapore, we should be focusing on the estate level. 

Example Of Estates And Regions: 
*  **Ang Mo Kio** (Estate)
*  **Chong Boon** (Region in Ang Mo Kio Estate)
*  **Cheng San** (Region in Ang Mo Kio Estate)
*  **Kebun Bahru** (Region in Ang Mo Kio Estate) 

---------------------------------------------------------

Another Example:
*   **Bedok** (Estate)
*   **Bayshore** (Region in Bedok Estate)
*   **Bedok North** (Region in Bedok Estate)
*   **Bedok South** (Region in Bedok Estate)
*   **Bedok Reservoir** (Region in Bedok Estate)

---------------------------------------------------------

For the dataset, we should only keep the estates such as `Ang Mo Kio` and `Bedok` in our analysis and not the regions. This can be done by extracting the values in the column `Number` which contain the word **'Total'**.
Also, we can remove the first row containing the grand total population (not us in our analysis)






In [91]:
# Removing regions and keeping estates
singapore_resident_data = singapore_resident_data[singapore_resident_data["Number"].str.contains("Total")]

# Removing the first row of the data 
singapore_resident_data = singapore_resident_data.iloc[1:]

singapore_resident_data.head(10)

Number Total_Total Total_0_4 Total_5_9  \
1                Ang Mo Kio- Total      174770      6790      7660   
14                    Bedok- Total      289750     11690     13400   
23                   Bishan- Total       90700      3430      4330   
27                 Boon Lay- Total          30        na        na   
32              Bukit Batok- Total      139270      5510      6890   
42              Bukit Merah- Total      155840      7210      7480   
60            Bukit Panjang- Total      139030      7330      7630   
68              Bukit Timah- Total       74470      3140      4660   
77  Central Water Catchment- Total          10        na        na   
79                   Changi- Total        2530       180       230   

   Total_10_14 Total_15_19 Total_20_24 Total_25_29 Total_30_34 Total_35_39  \
1         8290        9320       10310       11170       12250       13070   
14       14750       16930       19450       19860       19270       20850   
23        4710        5520        6860        6460        5720        6000   
27          na          na          na          na          na          na   
32        7970        9240       10070       10460        9760        9940   
42        6640        6860        7700        9340       11070       12840   
60        8460        9420       10400       10980       10690       10010   
68        4810        4820        4980        4310        3770        4940   
77          na          na          na          na          na          na   
79         220         150         130         100         170         240   

    ... Females_40_44 Females_45_49 Females_50_54 Females_55_59 Females_60_64  \
1   ...          7100          6660          7010          7120          6940   
14  ...         11660         10860         11870         11880         10570   
23  ...          3750          3590          4070          3960          3230   
27  ...            na            na            na            na            na   
32  ...          5680          5720          6180          5850          4610   
42  ...          6470          5540          5530          5780          5700   
60  ...          5770          5310          5880          5150          3860   
68  ...          3580          3290          3060          2780          2340   
77  ...            na            na            na            na            na   
79  ...           140           100            60            50            40   

   Females_65_69 Females_70_74 Females_75_79 Females_80_84 Females_85andOver  
1           5810          3600          2800          1910              1520  
14          8570          4880          4060          2840              2700  
23          2520          1460          1180           840               770  
27            na            na            na            na                na  
32          2980          1600          1170           870               720  
42          5440          3580          3110          2070              1980  
60          2560          1460          1120           780               680  
68          2010          1240          1030           660               630  
77            na            na            na            na                na  
79            40            30            10            20                20  

[10 rows x 58 columns]

##### **Step 2: Renaming The Values In The Number Column To The Estate Name**

Now it is time to rename the values in the Number column to the estate name. For instance, "Bedok- Total" will become "Bedok". This can be done simply by using the `.str.replace()` method. 

In [94]:
# Remove the '- Total' syntax from the Number column
singapore_resident_data["Number"] = singapore_resident_data["Number"].str.replace("- Total", "", regex = False)

singapore_resident_data.head(10)

Number Total_Total Total_0_4 Total_5_9 Total_10_14  \
1                Ang Mo Kio      174770      6790      7660        8290   
14                    Bedok      289750     11690     13400       14750   
23                   Bishan       90700      3430      4330        4710   
27                 Boon Lay          30        na        na          na   
32              Bukit Batok      139270      5510      6890        7970   
42              Bukit Merah      155840      7210      7480        6640   
60            Bukit Panjang      139030      7330      7630        8460   
68              Bukit Timah       74470      3140      4660        4810   
77  Central Water Catchment          10        na        na          na   
79                   Changi        2530       180       230         220   

   Total_15_19 Total_20_24 Total_25_29 Total_30_34 Total_35_39  ...  \
1         9320       10310       11170       12250       13070  ...   
14       16930       19450       19860       19270       20850  ...   
23        5520        6860        6460        5720        6000  ...   
27          na          na          na          na          na  ...   
32        9240       10070       10460        9760        9940  ...   
42        6860        7700        9340       11070       12840  ...   
60        9420       10400       10980       10690       10010  ...   
68        4820        4980        4310        3770        4940  ...   
77          na          na          na          na          na  ...   
79         150         130         100         170         240  ...   

   Females_40_44 Females_45_49 Females_50_54 Females_55_59 Females_60_64  \
1           7100          6660          7010          7120          6940   
14         11660         10860         11870         11880         10570   
23          3750          3590          4070          3960          3230   
27            na            na            na            na            na   
32          5680          5720          6180          5850          4610   
42          6470          5540          5530          5780          5700   
60          5770          5310          5880          5150          3860   
68          3580          3290          3060          2780          2340   
77            na            na            na            na            na   
79           140           100            60            50            40   

   Females_65_69 Females_70_74 Females_75_79 Females_80_84 Females_85andOver  
1           5810          3600          2800          1910              1520  
14          8570          4880          4060          2840              2700  
23          2520          1460          1180           840               770  
27            na            na            na            na                na  
32          2980          1600          1170           870               720  
42          5440          3580          3110          2070              1980  
60          2560          1460          1120           780               680  
68          2010          1240          1030           660               630  
77            na            na            na            na                na  
79            40            30            10            20                20  

[10 rows x 58 columns]

##### **Step 2: Handling NA Values In The Dataset**

From the results obtained earlier, we observe that there are a couple of columns that contain **na** values for the resident population. These include estates such as **Boon Lay** and **Central Water Catchment**. These na values make sense as these are estates where no one lives. 

For Example: 

**Boon Lay** -> Industrial Estate, no residential population

**Central Water Catchment** -> Pure Jungle, no inhabitants

Hence, it is best to simply remove these estates from the datasets as we are focusing on estates with residential population for accurate visualizations subsequently. 

**NOTE: The NA values are represented by na in text form, not NA/NaN. Hence, we cannot simply use the `.dropna()` method. We have to check if any row value contains na and then drop the row if there is at least one na value. However, this will exclude some estates with very low populations as well where some categories have na values and other categories have true figures.** 

In [97]:
# Removing na values from the dataset
singapore_resident_data = singapore_resident_data[
    ~singapore_resident_data.eq('na').any(axis=1)]

singapore_resident_data.head(10)

Number Total_Total Total_0_4 Total_5_9 Total_10_14 Total_15_19  \
1      Ang Mo Kio      174770      6790      7660        8290        9320   
14          Bedok      289750     11690     13400       14750       16930   
23         Bishan       90700      3430      4330        4710        5520   
32    Bukit Batok      139270      5510      6890        7970        9240   
42    Bukit Merah      155840      7210      7480        6640        6860   
60  Bukit Panjang      139030      7330      7630        8460        9420   
68    Bukit Timah       74470      3140      4660        4810        4820   
79         Changi        2530       180       230         220         150   
85  Choa Chu Kang      174330      7170      9340       11380       14820   
92       Clementi       91630      3910      4420        4480        4690   

   Total_20_24 Total_25_29 Total_30_34 Total_35_39  ... Females_40_44  \
1        10310       11170       12250       13070  ...          7100   
14       19450       19860       19270       20850  ...         11660   
23        6860        6460        5720        6000  ...          3750   
32       10070       10460        9760        9940  ...          5680   
42        7700        9340       11070       12840  ...          6470   
60       10400       10980       10690       10010  ...          5770   
68        4980        4310        3770        4940  ...          3580   
79         130         100         170         240  ...           140   
85       15230       11940       11540       11960  ...          7280   
92        4850        5820        6630        7310  ...          3940   

   Females_45_49 Females_50_54 Females_55_59 Females_60_64 Females_65_69  \
1           6660          7010          7120          6940          5810   
14         10860         11870         11880         10570          8570   
23          3590          4070          3960          3230          2520   
32          5720          6180          5850          4610          2980   
42          5540          5530          5780          5700          5440   
60          5310          5880          5150          3860          2560   
68          3290          3060          2780          2340          2010   
79           100            60            50            40            40   
85          8300          7830          6070          4180          2810   
92          3560          3480          3520          3740          3300   

   Females_70_74 Females_75_79 Females_80_84 Females_85andOver  
1           3600          2800          1910              1520  
14          4880          4060          2840              2700  
23          1460          1180           840               770  
32          1600          1170           870               720  
42          3580          3110          2070              1980  
60          1460          1120           780               680  
68          1240          1030           660               630  
79            30            10            20                20  
85          1600          1240           880               700  
92          1740          1380           870               780  

[10 rows x 58 columns]

Now we observe that non-residential estates such as Boon Lay and Central Water Catchment are removed from the dataset with no more na values.

##### **Step 3: Converting Absolute Numbers To Proportions**

For now, the data consists of absolute figures and numbers. This is not ideal as every estate has a different population and it will not be fair when we compare estates using absolute numbers. Some estates such as **Changi** have low population (2530 people only) while other estates such as **Bedok** have high population (289750 people). Hence, it is expected that Bedok will have many more people in each category compared to Changi. However, comparing each category in percentage form will solve the difference in population volume issue. 

In [101]:
# Convert all columns from the 3rd column onwards to numeric
singapore_resident_data.iloc[:, 2:] = singapore_resident_data.iloc[:, 2:].apply(
    pd.to_numeric, errors='coerce')

# Convert the total column (2nd column) to numeric
singapore_resident_data.iloc[:, 1] = pd.to_numeric(
    singapore_resident_data.iloc[:, 1], errors='coerce')

# Divide each age-group column by the total population (row-wise)
for col in singapore_resident_data.columns[2:]:
    singapore_resident_data[col] = (
        singapore_resident_data[col] / singapore_resident_data.iloc[:, 1])

# Multiply all values by 100 to get percentage form
singapore_resident_data.iloc[:, 2:] = singapore_resident_data.iloc[:, 2:] * 100

singapore_resident_data.head(10)

Number Total_Total Total_0_4 Total_5_9 Total_10_14 Total_15_19  \
1      Ang Mo Kio      174770  3.885106  4.382903    4.743377    5.332723   
14          Bedok      289750  4.034513  4.624676    5.090595    5.842968   
23         Bishan       90700  3.781698   4.77398    5.192944    6.085998   
32    Bukit Batok      139270  3.956344  4.947225    5.722697    6.634595   
42    Bukit Merah      155840   4.62654  4.799795     4.26078    4.401951   
60  Bukit Panjang      139030  5.272243  5.488024    6.085018    6.775516   
68    Bukit Timah       74470  4.216463  6.257553    6.458977    6.472405   
79         Changi        2530  7.114625  9.090909    8.695652    5.928854   
85  Choa Chu Kang      174330  4.112889  5.357655    6.527849    8.501119   
92       Clementi       91630  4.267161  4.823748    4.889228    5.118411   

   Total_20_24 Total_25_29 Total_30_34 Total_35_39  ... Females_40_44  \
1     5.899182    6.391257    7.009212      7.4784  ...      4.062482   
14    6.712683    6.854185    6.650561    7.195858  ...      4.024159   
23    7.563396    7.122381    6.306505    6.615215  ...      4.134509   
32    7.230559    7.510591     7.00797    7.137215  ...      4.078409   
42    4.940965    5.993326    7.103439     8.23922  ...      4.151694   
60      7.4804    7.897576    7.688988    7.199885  ...      4.150183   
68    6.687257    5.787565    5.062441    6.633544  ...      4.807305   
79     5.13834    3.952569    6.719368    9.486166  ...      5.533597   
85    8.736305    6.849079    6.619629    6.860552  ...      4.175988   
92    5.293026    6.351632    7.235622    7.977737  ...      4.299902   

   Females_45_49 Females_50_54 Females_55_59 Females_60_64 Females_65_69  \
1       3.810723      4.010986      4.073926      3.970933      3.324369   
14      3.748059      4.096635      4.100086      3.647972      2.957722   
23      3.958104      4.487321      4.366042      3.561191       2.77839   
32       4.10713      4.437424      4.200474      3.310117      2.139729   
42      3.554928      3.548511      3.708932      3.657598       3.49076   
60       3.81932      4.229303      3.704236      2.776379      1.841329   
68      4.417886      4.109037      3.733047      3.142205      2.699073   
79      3.952569      2.371542      1.976285      1.581028      1.581028   
85      4.761085      4.491482      3.481902      2.397751      1.611886   
92       3.88519      3.797883      3.841537      4.081633      3.601441   

   Females_70_74 Females_75_79 Females_80_84 Females_85andOver  
1        2.05985      1.602106      1.092865          0.869714  
14      1.684211      1.401208      0.980155          0.931838  
23      1.609702      1.300992       0.92613          0.848953  
32      1.148848      0.840095      0.624686          0.516981  
42      2.297228      1.995637      1.328285          1.270534  
60      1.050133      0.805582       0.56103          0.489103  
68        1.6651      1.383107      0.886263          0.845978  
79      1.185771      0.395257      0.790514          0.790514  
85        0.9178      0.711295       0.50479          0.401537  
92      1.898941      1.506057      0.949471           0.85125  

[10 rows x 58 columns]

##### **Step 4: Removing Unnecessary Columns**

The Unnecessary Columns Include: 
*   **Total_Total**: Still in absolute figures, inaccurate comparison and potentially misleading and confusing if kept.
*   **Total_0_4**, **Total_5_9**, **Total_10_14** etc: Redundant becuase we are interested in determining proportions by gender and age group for each estate, not the combined figures. 

The other remaining columns are stll relevant. The irrelevant columns corresponds from column 2 to column 21 of the dataset, as well as the column `Females_Total`.

In [104]:
# Remove the irrelevant columns of the dataset
singapore_resident_data = singapore_resident_data.drop(
    singapore_resident_data.columns[1:21], axis = 1)

singapore_resident_data = singapore_resident_data.drop(columns=['Females_Total'])

singapore_resident_data.head(10)

Number Males_0_4 Males_5_9 Males_10_14 Males_15_19 Males_20_24  \
1      Ang Mo Kio  1.991188  2.220061    2.403158    2.706414    2.935286   
14          Bedok  2.046592  2.329594    2.574633    3.030198     3.39258   
23         Bishan  1.984564  2.524807    2.745314    3.219405    3.958104   
32    Bukit Batok  2.024844  2.477203    2.879299    3.367559     3.62605   
42    Bukit Merah  2.361396  2.412731    2.194559    2.200975    2.457649   
60  Bukit Panjang  2.718838  2.769187    3.121628    3.452492     3.84809   
68    Bukit Timah   2.04109  2.712502    3.021351    3.048207    3.115348   
79         Changi  3.952569  4.743083    4.347826    3.162055    2.766798   
85  Choa Chu Kang  2.082258  2.776344    3.269661    4.399702    4.480009   
92       Clementi  2.149951  2.466441    2.466441    2.586489    2.641056   

   Males_25_29 Males_30_34 Males_35_39 Males_40_44  ... Females_40_44  \
1     3.078332    3.318647    3.536076    3.782114  ...      4.062482   
14    3.409836    3.175151     3.36497    3.748059  ...      4.024159   
23     3.53914    2.965821    3.098126    3.660419  ...      4.134509   
32    3.748115    3.331658     3.38192    3.805558  ...      4.078409   
42    2.906828    3.343172    3.901437     4.01694  ...      4.151694   
60    3.862476    3.797741    3.466878    3.891246  ...      4.150183   
68    2.645361    2.255942    2.913925    4.095609  ...      4.807305   
79    1.976285    2.766798    4.743083    5.928854  ...      5.533597   
85    3.487638    3.172145    3.252452    3.688407  ...      4.175988   
92    3.044854    3.470479    3.743316    3.983412  ...      4.299902   

   Females_45_49 Females_50_54 Females_55_59 Females_60_64 Females_65_69  \
1       3.810723      4.010986      4.073926      3.970933      3.324369   
14      3.748059      4.096635      4.100086      3.647972      2.957722   
23      3.958104      4.487321      4.366042      3.561191       2.77839   
32       4.10713      4.437424      4.200474      3.310117      2.139729   
42      3.554928      3.548511      3.708932      3.657598       3.49076   
60       3.81932      4.229303      3.704236      2.776379      1.841329   
68      4.417886      4.109037      3.733047      3.142205      2.699073   
79      3.952569      2.371542      1.976285      1.581028      1.581028   
85      4.761085      4.491482      3.481902      2.397751      1.611886   
92       3.88519      3.797883      3.841537      4.081633      3.601441   

   Females_70_74 Females_75_79 Females_80_84 Females_85andOver  
1        2.05985      1.602106      1.092865          0.869714  
14      1.684211      1.401208      0.980155          0.931838  
23      1.609702      1.300992       0.92613          0.848953  
32      1.148848      0.840095      0.624686          0.516981  
42      2.297228      1.995637      1.328285          1.270534  
60      1.050133      0.805582       0.56103          0.489103  
68        1.6651      1.383107      0.886263          0.845978  
79      1.185771      0.395257      0.790514          0.790514  
85        0.9178      0.711295       0.50479          0.401537  
92      1.898941      1.506057      0.949471           0.85125  

[10 rows x 37 columns]

##### **Step 5: Rename Columns To More Readable And Understandable Formats**

For now, the column names Number, Males_0_4 etc. are either confusing or vague. We can rename the columns to a more understandable and clear names, such as: 

*   Number -> Estate
*   Males_0_4 -> Males Age (<5)
*   Males_5_9 -> Males Age (5-9)
*   Females_0_4 -> Females Age (<5)
*   Females_85andOver -> Females Age (>=85)

In [107]:
# Renaming the useful columns to a more readable format
singapore_resident_data.rename(columns = {'Number': 'Estate', 
                                          'Males_0_4': 'Males Age (<5)',
                                          'Males_5_9': 'Males Age (5-9)',
                                          'Males_10_14': 'Males Age (10-14)',
                                          'Males_15_19': 'Males Age (15-19)',
                                          'Males_20_24': 'Males Age (20-24)',
                                          'Males_25_29': 'Males Age (25-29)',
                                          'Males_30_34': 'Males Age (30-34)',
                                          'Males_35_39': 'Males Age (35-39)',
                                          'Males_40_44': 'Males Age (40-44)',
                                          'Males_45_49': 'Males Age (45-49)',
                                          'Males_50_54': 'Males Age (50-54)',
                                          'Males_55_59': 'Males Age (55-59)',
                                          'Males_60_64': 'Males Age (60-64)',
                                          'Males_65_69': 'Males Age (65-69)',
                                          'Males_70_74': 'Males Age (70-74)',
                                          'Males_75_79': 'Males Age (75-79)',
                                          'Males_80_84': 'Males Age (80-84)',
                                          'Males_85andOver': 'Males Age (>=85)',
                                          'Females_0_4': 'Females Age (<5)',
                                          'Females_5_9': 'Females Age (5-9)',
                                          'Females_10_14': 'Females Age (10-14)',
                                          'Females_15_19': 'Females Age (15-19)',
                                          'Females_20_24': 'Females Age (20-24)',
                                          'Females_25_29': 'Females Age (25-29)',
                                          'Females_30_34': 'Females Age (30-34)',
                                          'Females_35_39': 'Females Age (35-39)',
                                          'Females_40_44': 'Females Age (40-44)',
                                          'Females_45_49': 'Females Age (45-49)',
                                          'Females_50_54': 'Females Age (50-54)',
                                          'Females_55_59': 'Females Age (55-59)',
                                          'Females_60_64': 'Females Age (60-64)',
                                          'Females_65_69': 'Females Age (65-69)',
                                          'Females_70_74': 'Females Age (70-74)',
                                          'Females_75_79': 'Females Age (75-79)',
                                          'Females_80_84': 'Females Age (80-84)',
                                          'Females_85andOver': 'Females Age (>=85)'},
                                           inplace = True)

singapore_resident_data.head(10)

Estate Males Age (<5) Males Age (5-9) Males Age (10-14)  \
1      Ang Mo Kio       1.991188        2.220061          2.403158   
14          Bedok       2.046592        2.329594          2.574633   
23         Bishan       1.984564        2.524807          2.745314   
32    Bukit Batok       2.024844        2.477203          2.879299   
42    Bukit Merah       2.361396        2.412731          2.194559   
60  Bukit Panjang       2.718838        2.769187          3.121628   
68    Bukit Timah        2.04109        2.712502          3.021351   
79         Changi       3.952569        4.743083          4.347826   
85  Choa Chu Kang       2.082258        2.776344          3.269661   
92       Clementi       2.149951        2.466441          2.466441   

   Males Age (15-19) Males Age (20-24) Males Age (25-29) Males Age (30-34)  \
1           2.706414          2.935286          3.078332          3.318647   
14          3.030198           3.39258          3.409836          3.175151   
23          3.219405          3.958104           3.53914          2.965821   
32          3.367559           3.62605          3.748115          3.331658   
42          2.200975          2.457649          2.906828          3.343172   
60          3.452492           3.84809          3.862476          3.797741   
68          3.048207          3.115348          2.645361          2.255942   
79          3.162055          2.766798          1.976285          2.766798   
85          4.399702          4.480009          3.487638          3.172145   
92          2.586489          2.641056          3.044854          3.470479   

   Males Age (35-39) Males Age (40-44)  ... Females Age (40-44)  \
1           3.536076          3.782114  ...            4.062482   
14           3.36497          3.748059  ...            4.024159   
23          3.098126          3.660419  ...            4.134509   
32           3.38192          3.805558  ...            4.078409   
42          3.901437           4.01694  ...            4.151694   
60          3.466878          3.891246  ...            4.150183   
68          2.913925          4.095609  ...            4.807305   
79          4.743083          5.928854  ...            5.533597   
85          3.252452          3.688407  ...            4.175988   
92          3.743316          3.983412  ...            4.299902   

   Females Age (45-49) Females Age (50-54) Females Age (55-59)  \
1             3.810723            4.010986            4.073926   
14            3.748059            4.096635            4.100086   
23            3.958104            4.487321            4.366042   
32             4.10713            4.437424            4.200474   
42            3.554928            3.548511            3.708932   
60             3.81932            4.229303            3.704236   
68            4.417886            4.109037            3.733047   
79            3.952569            2.371542            1.976285   
85            4.761085            4.491482            3.481902   
92             3.88519            3.797883            3.841537   

   Females Age (60-64) Females Age (65-69) Females Age (70-74)  \
1             3.970933            3.324369             2.05985   
14            3.647972            2.957722            1.684211   
23            3.561191             2.77839            1.609702   
32            3.310117            2.139729            1.148848   
42            3.657598             3.49076            2.297228   
60            2.776379            1.841329            1.050133   
68            3.142205            2.699073              1.6651   
79            1.581028            1.581028            1.185771   
85            2.397751            1.611886              0.9178   
92            4.081633            3.601441            1.898941   

   Females Age (75-79) Females Age (80-84) Females Age (>=85)  
1             1.602106            1.092865           0.869714  
14            1.401208            0.980155           0.931838

##### **Step 6: Rounding All Percentages To 2 Decimal Places**

In [110]:
# Rounding the average depression rates to 2 decimal places (for simplicity)
for col in singapore_resident_data.columns[1:]:
    singapore_resident_data[col] = pd.to_numeric(
        singapore_resident_data[col],
        errors='coerce'
    )

singapore_resident_data = singapore_resident_data.round(2)

singapore_resident_data.head(10)

Estate  Males Age (<5)  Males Age (5-9)  Males Age (10-14)  \
1      Ang Mo Kio            1.99             2.22               2.40   
14          Bedok            2.05             2.33               2.57   
23         Bishan            1.98             2.52               2.75   
32    Bukit Batok            2.02             2.48               2.88   
42    Bukit Merah            2.36             2.41               2.19   
60  Bukit Panjang            2.72             2.77               3.12   
68    Bukit Timah            2.04             2.71               3.02   
79         Changi            3.95             4.74               4.35   
85  Choa Chu Kang            2.08             2.78               3.27   
92       Clementi            2.15             2.47               2.47   

    Males Age (15-19)  Males Age (20-24)  Males Age (25-29)  \
1                2.71               2.94               3.08   
14               3.03               3.39               3.41   
23               3.22               3.96               3.54   
32               3.37               3.63               3.75   
42               2.20               2.46               2.91   
60               3.45               3.85               3.86   
68               3.05               3.12               2.65   
79               3.16               2.77               1.98   
85               4.40               4.48               3.49   
92               2.59               2.64               3.04   

    Males Age (30-34)  Males Age (35-39)  Males Age (40-44)  ...  \
1                3.32               3.54               3.78  ...   
14               3.18               3.36               3.75  ...   
23               2.97               3.10               3.66  ...   
32               3.33               3.38               3.81  ...   
42               3.34               3.90               4.02  ...   
60               3.80               3.47               3.89  ...   
68               2.26               2.91               4.10  ...   
79               2.77               4.74               5.93  ...   
85               3.17               3.25               3.69  ...   
92               3.47               3.74               3.98  ...   

    Females Age (40-44)  Females Age (45-49)  Females Age (50-54)  \
1                  4.06                 3.81                 4.01   
14                 4.02                 3.75                 4.10   
23                 4.13                 3.96                 4.49   
32                 4.08                 4.11                 4.44   
42                 4.15                 3.55                 3.55   
60                 4.15                 3.82                 4.23   
68                 4.81                 4.42                 4.11   
79                 5.53                 3.95                 2.37   
85                 4.18                 4.76                 4.49   
92                 4.30                 3.89                 3.80   

    Females Age (55-59)  Females Age (60-64)  Females Age (65-69)  \
1                  4.07                 3.97                 3.32   
14                 4.10                 3.65                 2.96   
23                 4.37                 3.56                 2.78   
32                 4.20                 3.31                 2.14   
42                 3.71                 3.66                 3.49   
60                 3.70                 2.78                 1.84   
68                 3.73                 3.14                 2.70   
79                 1.98                 1.58                 1.58   
85                 3.48                 2.40                 1.61   
92                 3.84                 4.08                 3.60   

    Females Age (70-74)  Females Age (75-79)  Females Age (80-84)  \
1                  2.06                 1.60                 1.09   
14                 1.68                 1.40                 0.98   
23                 1.61              

------------------------------------------------------------------------------------------------------------------